File: docs/sentence-similarity/query_corpus_embeddings.py
import sqlite3
from sentence_transformers import SentenceTransformer
from sklearn.metrics.pairwise import cosine_similarity
import numpy as np

# Initialize the Sentence Transformer model
# download it
# model = SentenceTransformer('all-MiniLM-L6-v2')

# use local copy
model = SentenceTransformer('./all-MiniLM-L6-v2')


# Function to calculate cosine similarity
def calculate_cosine_similarity(vec1, vec2):
    return cosine_similarity([vec1], [vec2])[0][0]

# Function to query similar quotes
def find_similar_quotes(query, db_file, threshold=0.3):
    # Step 1: Embed the query
    query_embedding = model.encode(query)

    # Step 2: Connect to the database
    conn = sqlite3.connect(db_file)
    cursor = conn.cursor()

    # Step 3: Retrieve all rows from the database
    cursor.execute("SELECT text, embedding FROM Quotes")
    rows = cursor.fetchall()

    results = []
    for text, embedding_str in rows:
        # Convert the embedding string back to a numpy array
        embedding = np.array([float(x) for x in embedding_str.split(",")])

        # Calculate cosine similarity
        similarity = calculate_cosine_similarity(query_embedding, embedding)

        if similarity > threshold:
            results.append((text, similarity))

    # Sort results by similarity in descending order
    results.sort(key=lambda x: x[1], reverse=True)

    # Close the database connection
    conn.close()

    return results

# Example usage
if __name__ == "__main__":
    db_file = "quotes.db"  # Path to the database file
    query = "Every difficulty brings opportunity."
    threshold = 0.3

    # Find similar quotes
    similar_quotes = find_similar_quotes(query, db_file, threshold)

    # Display results
    print("Input Query:")
    print(query)
    print("\nSimilar Quotes:")
    for quote, score in similar_quotes:
        print(f"{quote} (Similarity: {score:.2f})")

File: docs/semantic-reasoning/query_corpus_embeddings.kt
import android.database.sqlite.SQLiteDatabase
import kotlin.math.sqrt

fun calculateCosineSimilarity(vec1: List<Float>, vec2: List<Float>): Float {
    val dotProduct = vec1.zip(vec2).sumOf { it.first * it.second }
    val magnitude = sqrt(vec1.sumOf { it * it }) * sqrt(vec2.sumOf { it * it })
    return if (magnitude == 0.0f) 0.0f else dotProduct / magnitude
}

fun querySimilarQuotes(dbPath: String, queryText: String, threshold: Float): List<Pair<String, Float>> {
    // Step 1: Open the database
    val db = SQLiteDatabase.openDatabase(dbPath, null, SQLiteDatabase.OPEN_READONLY)

    // Step 2: Compute the embedding for the query
    val queryEmbedding = getQueryEmbedding(queryText) // Implement your embedding logic here

    // Step 3: Query the database for all quotes and embeddings
    val cursor = db.rawQuery("SELECT text, embedding FROM Quotes", null)
    val results = mutableListOf<Pair<String, Float>>()

    while (cursor.moveToNext()) {
        val text = cursor.getString(0)
        val embeddingStr = cursor.getString(1)

        // Convert embedding string back to a list of floats
        val embedding = embeddingStr.split(",").map { it.toFloat() }

        // Calculate cosine similarity
        val similarity = calculateCosineSimilarity(queryEmbedding, embedding)

        if (similarity > threshold) {
            results.add(text to similarity)
        }
    }

    cursor.close()
    db.close()

    // Step 4: Sort and return results
    return results.sortedByDescending { it.second }
}

fun getQueryEmbedding(queryText: String): List<Float> {
    // Use TensorFlow Lite, DJL, or an API to generate the embedding for the query
    // Placeholder: Return a dummy embedding for demonstration purposes
    return listOf(0.1f, 0.2f, 0.3f, /*...*/)
}

// Example usage
fun main() {
    val dbPath = "path_to_quotes.db"
    val queryText = "Every difficulty brings opportunity."
    val threshold = 0.3f

    val similarQuotes = querySimilarQuotes(dbPath, queryText, threshold)
    println("Input Query: $queryText")
    println("Similar Quotes:")
    for ((quote, similarity) in similarQuotes) {
        println("$quote (Similarity: $similarity)")
    }
}
